docs/contributing/effects-tutorial.md

```diff
@@ -0,0 +1,634 @@
+          doTheProgram
+        else doTheProgram
+```
+
+Note that other than some different function names (`modify'` -> `modify` and
+`liftIO` -> `sendIO`), and a very different (but still similar) style of listing
+function constraints, `doTheProgram` didn't change at all.  Our main slightly
+changed, but only such that we had to be specific about types. 
+
+Again, we can see that our `m` in `doTheProgram` is generic at the call site,
+but in our main function, we can see that it monomorphizes to
+`ReaderC AppConfig (StateC SeenValues (WriterC [ValidValue] IO))`.
+That's almost identical to our monad transformer monomorphization, and that
+similarity is very intentional.  Much of our intution from monad transformers
+is transferrable to `fused-effects`-style monads, and vice versa.
+
+## Pieces of the Puzzle
+
+Let's look at the pieces of `fused-effects`, and compare their purposes. We'll
+use the `Reader` concept as an example:
+
+| Name | Purpose |
+| --- | --- |
+| Effects | GADT-style datatypes which serve as message formats for an abstract interface. |
+| Carriers | Concrete monads which provide a implementation of an effectful interface. |
+| `Algebra` typeclass | The interface by which concrete implementations are selected and used. |
+| `Member` typeclass | Expressing and proving the effects required for an operation.
+
+Note that `fused-effects` differs from `mtl` in two very distinct ways:
+
+- Using `Algebra` instead of functional dependencies means that we can have multiple effects of the same type
+
+## Fitting the Pieces Together
+
+Let's try to explain the usage of `fused-effects` as simply as possible, again
+using the reader example:
+
+We run `ask` within some monadic context (often do-notation), which returns the
+value `m r`. `r` is the type that the reader is holding for us, and `m` is
+_some_ monad bound by the constraint `Has (Reader r) sig m`. Within the
+function, we don't care what the monad is, as long as we can bind it with `>>=`,
+or `<-` in do-notation, which is the definition of a monad in haskell.
+
+However, to make any program useful, it must eventually be reducible to `IO ()`,
+and so we need a way to make that `m` a concrete monad that satisfies the
+constraint, and then a way of reducing that concrete monad to either `IO` or a
+pure base monad like `Identity`.
+
+The concrete monad is a "transformer" of our base monad (`fused-effects` calls
+these `Carriers`) , which we compose by stacking them on top of each other, and
+we can use a runner to remove it from the stack of monads.  If our base monad is
+`IO`, we can pass that back up to `main`.  If it's a pure base monad, like
+`Identity`, we can use `runIdentity` to simply remove that base monad and get an
+entirely pure value.
+
+When we apply a runner, like `runReader`, that function specifies a _particular_
+carrier monad and removes it from the stack of carriers. Let's take a look at
+some of the definitions needed to make a reader work:
```

Comment:
Might want to link straight to the reader source too, in case anyone wants to see it in full.